File: backend/src/main/java/mk/ukim/finki/imdbclone/service/domain/RatingService.java
package mk.ukim.finki.imdbclone.service.domain;

import mk.ukim.finki.imdbclone.model.domain.Rating;

import java.util.List;
import java.util.Optional;

/**
 * Service interface for Rating operations.
 * Handles upsert of ratings and keeps {@code Media.averageRating} in sync.
 */
public interface RatingService {

    /**
     * Create or update a rating for a media item by a user.
     * After saving, {@code Media.averageRating} is recalculated and persisted.
     *
     * @param userId  the ID of the user rating the media
     * @param mediaId the ID of the media being rated
     * @param rating  the rating value (1–10)
     * @return the saved or updated Rating
     */
    Rating rateMedia(Long userId, Long mediaId, Integer rating);

    /**
     * Delete a user's rating for a media item.
     * After deletion, {@code Media.averageRating} is recalculated and persisted.
     *
     * @param userId  the user's ID
     * @param mediaId the media's ID
     */
    void deleteRating(Long userId, Long mediaId);

    /**
     * Find a specific user's rating for a media item
     *
     * @param userId  the user's ID
     * @param mediaId the media's ID
     * @return Optional containing the rating if found
     */
    Optional<Rating> getRating(Long userId, Long mediaId);

    /**
     * Find all ratings for a specific media item
     *
     * @param mediaId the media's ID
     * @return List of ratings
     */
    List<Rating> getRatingsByMedia(Long mediaId);

    /**
     * Find all ratings submitted by a specific user
     *
     * @param userId the user's ID
     * @return List of ratings
     */
    List<Rating> getRatingsByUser(Long userId);

    /**
     * Calculate the current average rating for a media item
     *
     * @param mediaId the media's ID
     * @return the average rating, or null if no ratings exist
     */
    Double getAverageRating(Long mediaId);

    /**
     * Count the total number of ratings for a media item
     *
     * @param mediaId the media's ID
     * @return the number of ratings
     */
    Long getRatingCount(Long mediaId);
}

File: frontend/src/lib/ratingService.js
import api from "./api";
import * as Rating from "../models/rating.model";

/**
 * Fetches the user's rating for a specific media.
 * @param {number} userId
 * @param {number} mediaId
 * @returns {Promise<Rating.Rating|null>}
 */
export async function getUserRating(userId, mediaId) {
    try {
        const response = await api.get(`/api/ratings/by-user-media`, {
            params: { userId, mediaId }
        });
        return Rating.fromApi(response.data);
    } catch (err) {
        if (err.response && err.response.status === 404) {
            return null;
        }
        throw err;
    }
}

/**
 * Submits a rating for a media.
 * @param {Object} ratingDto
 * @returns {Promise<Rating.Rating>}
 */
export async function rateMedia(ratingDto) {
    const response = await api.post("/api/ratings/add", ratingDto);
    return Rating.fromApi(response.data);
}

/**
 * Deletes a user's rating for a media.
 * @param {number} userId
 * @param {number} mediaId
 */
export async function deleteRating(userId, mediaId) {
    await api.delete("/api/ratings/delete", {
        params: { userId, mediaId }
    });
}

/**
 * Gets the average rating for a media item.
 * @param {number} mediaId
 */
export async function getAverageRating(mediaId) {
    const response = await api.get(`/api/ratings/media/${mediaId}/average`);
    return response.data;
}

/**
 * Gets the rating count for a media item.
 * @param {number} mediaId
 */
export async function getRatingCount(mediaId) {
    const response = await api.get(`/api/ratings/media/${mediaId}/count`);
    return response.data;
}
